src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,260 @@
+double CMAES::Optimize(DecomposableFunctionType& function, arma::mat& iterate)
+{
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+
+  // Population size.
+  if (lambda == 0)
+    lambda = (4 + std::round(3 * std::log(iterate.n_elem))) * 10;
+
+  // Parent weights.
+  const size_t mu = std::round(lambda / 2);
+  arma::vec w = std::log(mu + 0.5) - arma::log(arma::conv_to<arma::vec>::from(
+      arma::linspace<arma::uvec>(1, mu, mu + 1)));
+  w /= arma::sum(w);
+
+  // Number of effective solutions.
+  const double muEffective = 1 / arma::accu(arma::pow(w, 2));
+
+  // Step size control parameters.
+  arma::vec sigma(3);
+  sigma(0) = 0.3 * (upperBound - lowerBound);
+  const double cs = (muEffective + 2.0) / (iterate.n_elem + muEffective + 5.0);
+  const double ds = 1 + cs + 2 * std::max(std::sqrt((muEffective - 1) /
+      (iterate.n_elem + 1)) - 1, 0.0);
+  const double enn = std::sqrt(iterate.n_elem) * (1.0 - 1.0 /
+      (4.0 * iterate.n_elem) + 1.0 / (21 * std::pow(iterate.n_elem, 2)));
+
+  // Covariance update parameters.
+  // Cumulation for distribution.
+  const double cc = (4.0 + muEffective / iterate.n_elem) /
+      (4 + iterate.n_elem + 2 * muEffective / iterate.n_elem);
+  const double h = (1.4 + 2.0 / (iterate.n_elem + 1.0)) * enn;
+
+
+  const double c1 = 2 / (std::pow(iterate.n_elem + 1.3, 2) + muEffective);
+  const double alphaMu = 2;
+  const double cmu = std::min(1 - c1, alphaMu * (muEffective - 2 + 1 /
+      muEffective) / (std::pow(iterate.n_elem + 2, 2) +
+      alphaMu * muEffective / 2));
+
+  arma::cube mPosition(iterate.n_rows, iterate.n_cols, 3);
+  mPosition.slice(0) = lowerBound + arma::randu(
+      iterate.n_rows, iterate.n_cols) * (upperBound - lowerBound);
+
+  arma::mat step = arma::zeros(iterate.n_rows, iterate.n_cols);
+
+  double currentObjective = function.Evaluate(mPosition.slice(0), 0);
+  for (size_t f = 1; f < numFunctions; ++f)
+    currentObjective += function.Evaluate(mPosition.slice(0), f);
+
+  double overallObjective = currentObjective;
+  double lastObjective = DBL_MAX;
+
+  // Population parameters.
+  arma::cube pStep(iterate.n_rows, iterate.n_cols, lambda);
+  arma::cube pPosition(iterate.n_rows, iterate.n_cols, lambda);
+  arma::vec pObjective(lambda);
+  arma::cube ps = arma::zeros(iterate.n_rows, iterate.n_cols, 3);
+  arma::cube pc = arma::zeros(iterate.n_rows, iterate.n_cols, 3);
+  arma::cube C(iterate.n_elem, iterate.n_elem, 3);
+  C.slice(0).eye();
+
+  // Covariance matrix parameters.
+  arma::vec eigval;
+  arma::mat eigvec;
+  arma::vec eigvalZero = arma::zeros(iterate.n_elem);
+
+  // The current visitation order (sorted by population objectives).
+  arma::uvec idx = arma::linspace<arma::uvec>(0, lambda - 1, lambda);
+
+  // Now iterate!
+  for (size_t i = 1; i < maxIterations; ++i)
+  {
+    // To keep track of where we are.
+    const size_t idx0 = (i - 1) % 3;
+    const size_t idx1 = i % 3;
```

Comment:
I think that according to the algorithm you only ever need to look at time step `g` and `g + 1`; so maybe it is possible to reduce the number of slices in `C` to 2?  I am not certain, I have not checked completely.  I also don't know if it would make much difference to the running time or memory usage of the code anyway---just an idea I had.

Reply:
I'll change that, no need to allocate additional memory.